upgrade.sh

```diff
@@ -0,0 +1,261 @@
+lock_wait_start=$(date +%s)
+printed_status=false
+while fuser /var/lib/apt/lists/lock; do
+    if [ "${printed_status}" = false ]; then
+        echo "Waiting for another process to update package lists"
+        printed_status=true
+    fi
+    current_time=$(date +%s)
+    elapsed_time=$((current_time - lock_wait_start))
+    if [ "${elapsed_time}" -gt 30 ]; then break; fi
+    sleep 1
+done
+apt-get update
+
+# Wait up to 2 min for the dpkg lock to become available.
+lock_wait_start=$(date +%s)
+printed_status=false
+while fuser /var/lib/dpkg/lock; do
+    if [ "${printed_status}" = false ]; then
+        echo "Waiting for another process to update packages"
+        printed_status=true
+    fi
+    current_time=$(date +%s)
+    elapsed_time=$((current_time - lock_wait_start))
+    if [ "${elapsed_time}" -gt 120 ]; then break; fi
+    sleep 1
+done
+apt-get install -y git
+
+
+if [ -d /etc/appscale/certs ]; then
+    CURRENT_VERSION=$(cat /root/appscale/VERSION | grep -oE "[0-9]+\.[0-9]+\.[0-9]+")
+    if [ -z "${CURRENT_VERSION}" ]; then
+        echo "Cannot determine version of AppScale!"
+        exit 1
+    fi
+
+    # Make sure AppScale is not running.
+    MONIT=$(which monit)
+    if ps -o args -ax | grep -E '$[a-z/]+ruby[ a-zA-Z/-]+djinnServer.rb' > /dev/null ; then
+        echo "AppScale is still running: please stop it"
+        exit 1
+    elif echo ${MONIT} | grep local > /dev/null ; then
+        # AppScale is not running but there is a monit
+        # leftover from the custom install.
+        ${MONIT} quit
+    fi
+
+    # Let's keep a copy of the old config: we need to move it to avoid
+    # questions from dpkg.
+    if [ -e /etc/haproxy/haproxy.cfg ]; then
+        mv /etc/haproxy/haproxy.cfg /etc/haproxy/haproxy.cfg.appscale.old
+    fi
+
+    # Remove outdated appscale-controller and appscale-progenitor.
+    if version_ge 2.2.0 "${CURRENT_VERSION}"; then
+        rm -f /etc/init.d/appscale-controller
+        rm -f /etc/init.d/appscale-progenitor
+        update-rc.d -f appscale-progenitor remove || true
+    fi
+
+    # Remove control files we added before 1.14, and re-add the
+    # default ones.
+    if version_ge 1.14.0 "${CURRENT_VERSION}"; then
+        rm -f /etc/default/haproxy /etc/init.d/haproxy /etc/default/monit /etc/monitrc
+        if dpkg-query -l haproxy > /dev/null 2> /dev/null ; then
+            apt-get -o DPkg::Options::="--force-confmiss" --reinstall install haproxy
+        fi
+        if dpkg-query -l monit > /dev/null 2> /dev/null ; then
+            apt-get -o DPkg::Options::="--force-confmiss" --reinstall install monit
+        fi
+    fi
+
+    echo "Found AppScale version ${CURRENT_VERSION}. "\
+         "An upgrade to the ${GIT_TAG} version will be attempted in 5 seconds."
+    echo "Exit now (ctrl-c) if this is incorrect"
+    sleep 5
+else
+    echo "No previous AppScale installation was found. Installing it from scratch."
```

Comment:
Why do we have 2 scripts if this one can also do straight installation?

Reply:
After https://github.com/AppScale/appscale/pull/3140 (which is included into this branch) boostrap script will be split into flexible bootstrap (it will allow custom branches) and restricted upgrade (it will allow upgrade to tag only).